### PR TITLE
Seperate App to Web Server and Device Server

### LINKED ISSUE #58: [API] Setup two Express servers
As per our last meeting, we will be splitting the device-facing and webapp-facing routes into two Express servers. This will involve:
1. Modify `app.js` to start up two express servers. This means calling `express()` twice and having two different `app` variables. We can call one `deviceApp` and one `webApp`. For now, they should both have the same `.use` calls.
2. Modify `util/router.js` to take a second argument that tells it from which subfolder in `/routes` to load routes, so that when you call it for each app in `app.js`, you can do it like `require('./util/router').boot(webApp, 'backend')` or `require('./util/router').boot(deviceApp, 'api')`, so each server will have different routes. This will use your renamed folder names.

### PR DESCRIPTION
**Pull request by Lepabryce**
Separate the App into two separate express servers as was discussed in a previous meeting. Both servers will use HTTPS and require certificates when in production. I went through all the route files to make sure they export the correct app variable. Closes #58

### CODE REVIEW

**Comment by harsimrat99** [comment]
@matt8100 could you please review Bryce's updates in this PR? Thank you.

File: routes/api/auth/GET-cert.js [hunk 1]
```diff
@@ -3,8 +3,8 @@ const Archiver = require('archiver');
 const jwtVerify = require('../../../middleware/jwtVerify');
 const genCert = require('../../../util/genCert');
 
-module.exports = (app) => {
-  app.get('/api/auth/cert', jwtVerify, (req, res) => {
+module.exports = (webApp) => {
```
Comment (matt8100) [inline routes/api/auth/GET-cert.js]:
```suggestion
module.exports = (app) => {
```
Unnecessary rename

File: routes/api/auth/GET-cert.js [hunk 2]
```diff
@@ -3,8 +3,8 @@ const Archiver = require('archiver');
...
+  webApp.get('/api/auth/cert', jwtVerify, (req, res) => {
```
Comment (matt8100) [inline routes/api/auth/GET-cert.js]:
```suggestion
  app.get('/api/auth/cert', jwtVerify, (req, res) => {
```
Unnecessary rename

File: routes/api/auth/POST-login.js [hunk 3]
```diff
@@ -1,8 +1,8 @@
 const login = require('../../../middleware/login');
 const jwtSign = require('../../../middleware/jwtSign');
 
-module.exports = (app) => {
-  app.post('/api/auth/login', login, jwtSign, (req, res) => {
+module.exports = (webApp) => {
```
Comment (matt8100) [inline routes/api/auth/POST-login.js]:
```suggestion
module.exports = (app) => {
```
Unnecessary rename

File: routes/api/auth/POST-login.js [hunk 4]
```diff
@@ -1,8 +1,8 @@
...
+  webApp.post('/api/auth/login', login, jwtSign, (req, res) => {
```
Comment (matt8100) [inline routes/api/auth/POST-login.js]:
```suggestion
  app.post('/api/auth/login', login, jwtSign, (req, res) => {
```
Unnecessary rename

File: routes/api/data/POST-blob.js [hunk 5]
```diff
@@ -5,8 +5,8 @@ const jwtVerify = require('../../../middleware/jwtVerify');
 const s3Client = require('../../../util/s3'); // configure s3 client
 const BlobLog = require('../../../models/blobLog');
 
-module.exports = (app) => {
-  app.post('/api/data/blob', jwtVerify, (req, res, next) => {
+module.exports = (webApp) => {
```
Comment (matt8100) [inline routes/api/data/POST-blob.js]:
```suggestion
module.exports = (app) => {
```
Unnecessary rename

File: routes/api/data/POST-blob.js [hunk 6]
```diff
@@ -5,8 +5,8 @@ const jwtVerify = require('../../../middleware/jwtVerify');
...
+  webApp.post('/api/data/blob', jwtVerify, (req, res, next) => {
```
Comment (matt8100) [inline routes/api/data/POST-blob.js]:
```suggestion
  app.post('/api/data/blob', jwtVerify, (req, res, next) => {
```
Unnecessary rename

File: routes/api/data/POST-new.js [hunk 7]
```diff
@@ -2,8 +2,8 @@ const jwtVerify = require('../../../middleware/jwtVerify');
 const Data = require('../../../models/data');
 const Device = require('../../../models/device');
 
-module.exports = (app) => {
-  app.post('/api/data/new', jwtVerify, (req, res) => {
+module.exports = (webApp) => {
```
Comment (matt8100) [inline routes/api/data/POST-new.js]:
```suggestion
module.exports = (app) => {
```
Unnecessary rename

File: routes/api/data/POST-new.js [hunk 8]
```diff
@@ -2,8 +2,8 @@ const jwtVerify = require('../../../middleware/jwtVerify');
...
+  webApp.post('/api/data/new', jwtVerify, (req, res) => {
```
Comment (matt8100) [inline routes/api/data/POST-new.js]:
```suggestion
  app.post('/api/data/new', jwtVerify, (req, res) => {
```
Unnecessary rename

File: routes/device/GET-info.js [hunk 9]
```diff
@@ -1,7 +1,7 @@
 const Device = require('../../models/device');
 
-module.exports = (app) => {
-  app.get('/device/info', (req, res) => {
+module.exports = (deviceApp) => {
```
Comment (matt8100) [inline routes/device/GET-info.js]:
```suggestion
module.exports = (app) => {
```
Unnecessary rename

File: routes/device/GET-info.js [hunk 10]
```diff
@@ -1,7 +1,7 @@
...
+  deviceApp.get('/device/info', (req, res) => {
```
Comment (matt8100) [inline routes/device/GET-info.js]:
```suggestion
  app.get('/device/info', (req, res) => {
```
Unnecessary rename

File: routes/device/POST-cert.js [hunk 11]
```diff
@@ -1,8 +1,8 @@
 const login = require('../../middleware/login');
 const genCert = require('../../util/genCert');
 
-module.exports = (app) => {
-  app.post('/device/cert', login, (req, res) => {
+module.exports = (deviceApp) => {
```
Comment (matt8100) [inline routes/device/POST-cert.js]:
```suggestion
module.exports = (app) => {
```
Unnecessary rename

File: routes/device/POST-cert.js [hunk 12]
```diff
@@ -1,8 +1,8 @@
...
+  deviceApp.post('/device/cert', login, (req, res) => {
```
Comment (matt8100) [inline routes/device/POST-cert.js]:
```suggestion
  app.post('/device/cert', login, (req, res) => {
```
Unnecessary rename

File: routes/device/POST-register.js [hunk 13]
```diff
@@ -1,8 +1,8 @@
 const jwtVerify = require('../../middleware/jwtVerify');
 const Device = require('../../models/device');
 
-module.exports = (app) => {
-  app.post('/device/register', jwtVerify, (req, res) => {
+module.exports = (deviceApp) => {
```
Comment (matt8100) [inline routes/device/POST-register.js]:
```suggestion
module.exports = (app) => {
```
Unnecessary rename

File: routes/device/POST-register.js [hunk 14]
```diff
@@ -1,8 +1,8 @@
...
+  deviceApp.post('/device/register', jwtVerify, (req, res) => {
```
Comment (matt8100) [inline routes/device/POST-register.js]:
```suggestion
  app.post('/device/register', jwtVerify, (req, res) => {
```
Unnecessary rename

File: routes/device/POST-unregister.js [hunk 15]
```diff
@@ -1,8 +1,8 @@
 const jwtVerify = require('../../middleware/jwtVerify');
 const Device = require('../../models/device');
 
-module.exports = (app) => {
-  app.post('/device/unregister', jwtVerify, (req, res) => {
+module.exports = (deviceApp) => {
```
Comment (matt8100) [inline routes/device/POST-unregister.js]:
```suggestion
module.exports = (app) => {
```
Unnecessary rename

File: routes/device/POST-unregister.js [hunk 16]
```diff
@@ -1,8 +1,8 @@
...
+  deviceApp.post('/device/unregister', jwtVerify, (req, res) => {
```
Comment (matt8100) [inline routes/device/POST-unregister.js]:
```suggestion
  app.post('/device/unregister', jwtVerify, (req, res) => {
```
Unnecessary rename

File: routes/device/data/GET-getData.js [hunk 17]
```diff
@@ -1,8 +1,8 @@
 const jwtVerify = require('../../../middleware/jwtVerify');
 const Data = require('../../../models/data');
 
-module.exports = (app) => {
-  app.get('/device/data/getData', jwtVerify, (req, res) => {
+module.exports = (deviceApp) => {
```
Comment (matt8100) [inline routes/device/data/GET-getData.js]:
```suggestion
module.exports = (app) => {
```
Unnecessary rename

File: routes/device/data/GET-getData.js [hunk 18]
```diff
@@ -1,8 +1,8 @@
...
+  deviceApp.get('/device/data/getData', jwtVerify, (req, res) => {
```
Comment (matt8100) [inline routes/device/data/GET-getData.js]:
```suggestion
  app.get('/device/data/getData', jwtVerify, (req, res) => {
```
Unnecessary rename

File: app.js [hunk 19]
```diff
@@ -1,21 +1,54 @@
...
+const webApp = express();
+webApp.use(express.urlencoded({ extended: true }));
+webApp.use(express.json());
+webApp.use(cookieParser());
+if (process.env.PRODUCTION) {
+  const deviceAppOptions = {
+    key: fs.readFileSync(process.env.WEB_SERVER_KEY_PATH),
+    cert: fs.readFileSync(process.env.WEB_SERVER_CERT_PATH),
+    ca: fs.readFileSync(process.env.WEB_SERVER_CA_PATH),
+    requestCert: true,
```
Comment (matt8100) [inline app.js]:
```suggestion
```
Webapp-facing server does not require signed certificate authorization

File: app.js [hunk 20]
```diff
@@ -1,21 +1,54 @@
...
+  const webAppOptions = {
+    key: fs.readFileSync(process.env.DEVICE_SERVER_KEY_PATH),
+    cert: fs.readFileSync(process.env.DEVICE_SERVER_CERT_PATH),
+    ca: fs.readFileSync(process.env.DEVICE_SERVER_CA_PATH),
+    requestCert: true,
+  };
+  // Dynamic route loading
+  require('./util/router').boot(deviceApp, 'device');
+  https.createServer(deviceAppOptions, deviceApp).listen(process.env.PORT || 3000, function listen() {
+    console.log(`Express server for device utilities started and listening on port ${this.address().port}`);
```
Comment (matt8100) [inline app.js]:
```suggestion
    console.log(`Device HTTPS server started and listening on port ${this.address().port}`);
```
More meaningful startup message

File: app.js [hunk 21]
```diff
@@ -1,21 +1,54 @@
...
+  });
+  require('./util/router').boot(webApp, 'api');
+  https.createServer(webAppOptions, webApp).listen(process.env.WEBPORT || 3001, function listen() {
+    console.log(`Express server for web utilities started and listening on port ${this.address().port}`);
```
Comment (matt8100) [inline app.js]:
```suggestion
    console.log(`Webapp HTTPS server started and listening on port ${this.address().port}`);
```
More meaningful startup message

File: app.js [hunk 22]
```diff
@@ -1,21 +1,54 @@
...
+  });
+} else {
+  require('./util/router').boot(deviceApp, 'device');
+  deviceApp.listen(process.env.PORT || 3000, function listen() {
+    console.log(`Express server for device utilities started and listening on port ${this.address().port}`);
```
Comment (matt8100) [inline app.js]:
```suggestion
    console.log(`Device HTTP server started and listening on port ${this.address().port}`);
```
More meaningful startup message

File: app.js [hunk 23]
```diff
@@ -1,21 +1,54 @@
...
+  });
+  require('./util/router').boot(webApp, 'api');
+  webApp.listen(process.env.PORT || 3001, function listen() {
+    console.log(`Express server for web utilities started and listening on port ${this.address().port}`);
```
Comment (matt8100) [inline app.js]:
```suggestion
    console.log(`Webapp HTTP server started and listening on port ${this.address().port}`);
```
More meaningful startup message

File: app.js [hunk 24]
```diff
@@ -1,21 +1,54 @@
+/* eslint-disable global-require */
+/* eslint-disable max-len */
 require('./util/env').configure(); // configure the environment variables
 require('./util/mongodb'); // configure mongoose db connection
 require('./util/aedes'); // configure MQTT broker
 require('./util/s3'); // configure s3 client
+const https = require('https');
```
Comment (matt8100) [inline app.js]:
```suggestion
const https = (process.env.PRODUCTION) ? require('https') : {};
```
Do not require module unless needed

File: test/auth_cert_test.js [hunk 25]
```diff
@@ -5,7 +5,7 @@ const jwt = require('jsonwebtoken');
 const forge = require('node-forge');
 const fs = require('fs');
 
-const app = require('../app');
+const apps = require('../app');
```
Comment (matt8100) [inline test/auth_cert_test.js]:
```suggestion
const { webApp } = require('../app');
```
Use destructuring

File: test/auth_cert_test.js [hunk 26]
```diff
@@ -18,7 +18,7 @@ describe('/GET api/auth/cert', () => {
       algorithm: 'HS256',
       expiresIn: 60,
     });
-    chai.request(app)
+    chai.request(apps.webApp)
```
Comment (matt8100) [inline test/auth_cert_test.js]:
```suggestion
    chai.request(webApp)
```
Use destructuring

File: test/auth_login_test.js [hunk 27]
```diff
@@ -6,7 +6,7 @@ const jwt = require('jsonwebtoken');
 
 process.env.DB_NAME = 'test';
 
-const app = require('../app');
+const apps = require('../app');
```
Comment (matt8100) [inline test/auth_login_test.js]:
```suggestion
const { webApp } = require('../app');
```
Use destructuring

File: test/auth_login_test.js [hunk 28]
```diff
@@ -26,7 +26,7 @@ describe('/POST api/auth/login', () => {
   });
 
   it('it should return a valid auth cookie', (done) => {
-    chai.request(app)
+    chai.request(apps.webApp)
```
Comment (matt8100) [inline test/auth_login_test.js]:
```suggestion
    chai.request(webApp)
```
Use destructuring

File: test/data_blob_test.js [hunk 29]
```diff
@@ -8,7 +8,7 @@ const jwt = require('jsonwebtoken');
 const s3Client = require('../util/s3'); // configure s3 client
 const BlobLog = require('../models/blobLog');
 
-const app = require('../app');
+const apps = require('../app');
```
Comment (matt8100) [inline test/data_blob_test.js]:
```suggestion
const { webApp } = require('../app');
```
Use destructuring

File: test/data_blob_test.js [hunk 30]
```diff
@@ -38,7 +38,7 @@ describe('/POST api/data/blob', function test() {
       expiresIn: 60,
     });
 
-    chai.request(app)
+    chai.request(apps.webApp)
```
Comment (matt8100) [inline test/data_blob_test.js]:
```suggestion
    chai.request(webApp)
```
Use destructuring

File: test/data_getData_test.js [hunk 31]
```diff
@@ -3,7 +3,7 @@ const chaiHttp = require('chai-http');
 const faker = require('faker');
 const jwt = require('jsonwebtoken');
 
-const app = require('../app');
+const apps = require('../app');
```
Comment (matt8100) [inline test/data_getData_test.js]:
```suggestion
const { deviceApp } = require('../app');
```
Use destructuring

File: test/data_getData_test.js [hunk 32]
```diff
@@ -41,7 +41,7 @@ describe('/GET device/data/getData', function test() {
   });
 
   it('it should GET the data', (done) => {
-    chai.request(app)
+    chai.request(apps.deviceApp)
```
Comment (matt8100) [inline test/data_getData_test.js]:
```suggestion
    chai.request(deviceApp)
```
Use destructuring

File: test/data_new_test.js [hunk 33]
```diff
@@ -3,7 +3,7 @@ const jwt = require('jsonwebtoken');
 const chaiHttp = require('chai-http');
 const faker = require('faker');
 
-const app = require('../app');
+const apps = require('../app');
```
Comment (matt8100) [inline test/data_new_test.js]:
```suggestion
const { webApp } = require('../app');
```
Use destructuring

File: test/data_new_test.js [hunk 34]
```diff
@@ -35,7 +35,7 @@ describe('/POST api/data/new', () => {
   });
 
   it('it should POST the data', (done) => {
-    chai.request(app)
+    chai.request(apps.webApp)
```
Comment (matt8100) [inline test/data_new_test.js]:
```suggestion
    chai.request(webApp)
```
Use destructuring

File: test/device_cert_test.js [hunk 35]
```diff
@@ -5,7 +5,7 @@ const bcrypt = require('bcrypt');
 const forge = require('node-forge');
 const fs = require('fs');
 
-const app = require('../app');
+const apps = require('../app');
```
Comment (matt8100) [inline test/device_cert_test.js]:
```suggestion
const { deviceApp } = require('../app');
```
Use destructuring

File: test/device_cert_test.js [hunk 36]
```diff
@@ -25,7 +25,7 @@ describe('/POST device/cert', () => {
 
   it('it should authenticate the device and return a client certificate and key', (done) => {
     const caCert = forge.pki.certificateFromPem(fs.readFileSync(process.env.MQTT_SERVER_CA_PATH));
-    chai.request(app)
+    chai.request(apps.deviceApp)
```
Comment (matt8100) [inline test/device_cert_test.js]:
```suggestion
    chai.request(deviceApp)
```
Use destructuring

File: test/device_info_test.js [hunk 37]
```diff
@@ -2,7 +2,7 @@ const chai = require('chai');
 const chaiHttp = require('chai-http');
 const faker = require('faker');
 
-const app = require('../app');
+const apps = require('../app');
```
Comment (matt8100) [inline test/device_info_test.js]:
```suggestion
const { deviceApp } = require('../app');
```
Use destructuring

File: test/device_info_test.js [hunk 38]
```diff
@@ -21,7 +21,7 @@ describe('/GET device/info', () => {
   });
 
   it('it should GET the device information', (done) => {
-    chai.request(app)
+    chai.request(apps.deviceApp)
```
Comment (matt8100) [inline test/device_info_test.js]:
```suggestion
    chai.request(deviceApp)
```
Use destructuring

File: test/device_register_test.js [hunk 39]
```diff
@@ -3,23 +3,27 @@ const chaiHttp = require('chai-http');
 const faker = require('faker');
 const jwt = require('jsonwebtoken');
 
-const app = require('../app');
+const apps = require('../app');
```
Comment (matt8100) [inline test/device_register_test.js]:
```suggestion
const { deviceApp } = require('../app');
```
Use destructuring

File: test/device_register_test.js [hunk 40]
```diff
@@ -3,23 +3,27 @@ const chaiHttp = require('chai-http');
...
+      algorithm: 'HS256',
+      expiresIn: 60,
+    });
   });
 
   const tempMac = faker.internet.mac().replace(/:/g, '');
   const friendlyName = faker.internet.userName();
   it('it should POST the information', (done) => {
-    chai.request(app)
+    chai.request(apps.deviceApp)
```
Comment (matt8100) [inline test/device_register_test.js]:
```suggestion
    chai.request(deviceApp)
```
Use destructuring

File: test/device_unregister_test.js [hunk 41]
```diff
@@ -3,22 +3,22 @@ const chaiHttp = require('chai-http');
 const faker = require('faker');
 const jwt = require('jsonwebtoken');
 
-const app = require('../app');
+const apps = require('../app');
```
Comment (matt8100) [inline test/device_unregister_test.js]:
```suggestion
const { deviceApp } = require('../app');
```
Use destructuring

File: test/device_unregister_test.js [hunk 42]
```diff
@@ -27,7 +27,7 @@ describe('/POST device/unregister', () => {
   });
 
   it('it should use POST to remove the information', (done) => {
-    chai.request(app)
+    chai.request(apps.deviceApp)
```
Comment (matt8100) [inline test/device_unregister_test.js]:
```suggestion
    chai.request(deviceApp)
```
Use destructuring

**Review by harsimrat99 — APPROVED** [review]

**Pull request merged**